lib/internal/modules/esm/loader.js

```diff
@@ -166,7 +167,7 @@ class ESMLoader {
   ) {
     if (!ArrayIsArray(customLoaders)) customLoaders = [customLoaders];
 
-    for (let i = 0; i < customLoaders.length; i++) {
+    for (let i = 0, count = customLoaders.length; i < count; i++) {
```

Comment:
nit: in JS `length` is an own property of the array, it doesn't matter if you access it from a variable or the `length` property.
```suggestion
    for (let i = 0; i < customLoaders.length; i++) {
```

Reply:
Are you sure it doesn't matter? `arr.length` is a getter, computed JIT every time it's accessed, so it's usually much better for performance to cache it when assured it won't (or shouldn't) change. I believe that non-changing scenario applies here.

Reply:
It's not a getter, it's an own property. Your reasoning would be true for `TypedArray` though:
```js
Object.getOwnPropertyDescriptor([], 'length'); // { value: 0, ...}
Object.getOwnPropertyDescriptor(new Uint8Array(), 'length'); // undefined
```

Reply:
Hmm, I'm fairly certain I've read from an authoritative source that there is some expense to checking `arr.length` and repeated reads should be avoided when possible—and also saw some benchmarking that supported it. It has been a couple/few years since last I looked, so my info could also be outdated. I can do a bit of digging if you'd like to see something to back that up.

Do you have a particularly reason to request the proposed alternate, or is it a preference (or consistency) thing? I ask because your initial comment sounds like maybe it's a preference.

Reply:
I'd be interested of being shown wrong, because in https://github.com/nodejs/node/pull/38635 I'm putting some "best practices" code snippets, so if there's a performance concern, I'd like to know about it 😅  I am fairly confident this advice wasn't meant for JS arrays (maybe another language or another type of object?), the JS engine MUST know the length of an array at any time (because an array may be holy), but if you have info pointing otherwise please share.

> Do you have a particularly reason to request the proposed alternate, or is it a preference (or consistency) thing? I ask because your initial comment sounds like maybe it's a preference.

It was in the spirit of simplifying the code, if you have a personal preference for one or the other, go for it, I don't feel strongly either way.

Reply:
Unfortunately, most of the benchmarking was on jsperf, which is now dead. I set up a quick n dirty benchmark locally using [benchmark.js](https://github.com/bestiejs/benchmark.js)

<details>
<summary>looping over an array with 10k elements (and doing nothing)</summary>

```js
const arr = new Array(10000);

// uncached length, incrementing
for (let i = 0; i < arr.length; i++);

// cached length, incrementing
for (let i = 0, length = arr.length; i < length; i++);

// cached length, decrementing
for (let i = arr.length - 1; i > -1; i--);

```

</details>

and got:

> uncached length, incrementing x 295,816 ops/sec ±0.48% (95 runs sampled)
> cached length, incrementing x 296,260 ops/sec ±0.40% (97 runs sampled)
> cached length, decrementing x 294,305 ops/sec ±0.60% (97 runs sampled)
>
> fastest is: **cached length, incrementing**

The difference is fairly small, but cached length is consistently better and has the least variance.

Curious, because decrementing used to be far superior (at some point, it saved an under-the-hood out of bounds check at every iteration because it was already starting at the upper bound and the lower bound is known/always zero).

---

My personal preference is only based on performance. Aesthetically, they're basically identical, so I have no opinion if there's no performance benefit. Well, my personal preference is for…of because of convenience and consistency across data types (I don't have to update code when switching from one iterable to another); but that's not an option here.

Reply:
I'm happy to go with whichever. If the recommendation is already written up and roll out started, seems easier to align with that. I'll push a separate commit with it just in case.